godot-core/src/builtin/arrays.rs

```diff
@@ -135,15 +146,68 @@ impl Array {
         self.as_inner().resize(to_i64(size));
     }
 
-    // TODO: find out why this segfaults (even on an empty array, regardless of deep = true/false)
-    // /// Returns a deep copy of the array. All nested arrays and dictionaries are duplicated and
-    // /// will not be shared with the original array. Note that any `Object`-derived elements will
-    // /// still be shallow copied.
-    // ///
-    // /// To create a shallow copy, use `clone()` instead.
-    // pub fn duplicate_deep(&self) -> Self {
-    //     self.as_inner().duplicate(true)
-    // }
+    /// Returns a shallow copy of the array. All array elements are copied, but any reference types
+    /// (such as `Array`, `Dictionary` and `Object`) will still refer to the same value.
+    ///
+    /// To create a deep copy, use [`duplicate_deep()`] instead. To create a new reference to the
+    /// same array data, use `clone()`.
+    pub fn duplicate_shallow(&self) -> Self {
+        self.as_inner().duplicate(false)
+    }
+
+    /// Returns a deep copy of the array. All nested arrays and dictionaries are duplicated and
+    /// will not be shared with the original array. Note that any `Object`-derived elements will
+    /// still be shallow copied.
+    ///
+    /// To create a shallow copy, use [`duplicate_shallow()`] instead. To create a new reference to
+    /// the same array data, use `clone()`.
+    pub fn duplicate_deep(&self) -> Self {
+        self.as_inner().duplicate(true)
+    }
+
+    /// Returns the slice of the `Array`, from `begin` (inclusive) to `end` (exclusive), as a new
+    /// `Array`.
+    ///
+    /// The values of `begin` and `end` will be clamped to the array size.
+    ///
+    /// If specified, `step` is the relative index between source elements. It can be negative,
+    /// in which case `begin` must be higher than `end`. For example,
+    /// `Array::from(&[0, 1, 2, 3, 4, 5]).slice(5, 1, -2)` returns `[5, 3]`.
+    ///
+    /// Array elements are copied to the slice, but any reference types (such as `Array`,
+    /// `Dictionary` and `Object`) will still refer to the same value. To create a deep copy, use
+    /// [`slice_deep()`] instead.
+    pub fn slice_shallow(&self, begin: usize, end: usize, step: Option<isize>) -> Self {
+        let len = self.len();
+        let begin = begin.min(len);
+        let end = end.min(len);
+        let step = step.unwrap_or(1);
+        assert!(step != 0);
```

Comment:
`assert_ne!` and maybe at the beginning of the function, so that invariant checks are immediately visible?
```rs
assert_ne!(step, Some(0));
```

Reply:
Done.